bookserver/routers/rslogging.py

```diff
@@ -179,3 +188,145 @@ async def runlog(request: Request, response: Response, data: LogRunIncoming):
 async def same_class(user1: AuthUserValidator, user2: str) -> bool:
     u2 = await fetch_user(user2)
     return user1.course_id == u2.course_id
+
+
+# completion tables
+# =================
+#
+# This section contains implementations of endpoints for tracking progress
+
+
+# updatelastpage
+# --------------
+@router.post("updatelastpage")
+async def updatelastpage(request: Request, request_data: LastPageIncoming):
+    if request_data.last_page_url is None:
+        return  # todo:  log request_data, request.args and request.env.path_info
+
+    lpd = LastPageData(**request_data.dict())
+
+    lpd.last_page_chapter = request_data.last_page_url.split("/")[-2]
+    lpd.last_page_subchapter = ".".join(
+        request_data.last_page_url.split("/")[-1].split(".")[:-1]
+    )
+
+    if request.state.user:
```

Comment:
I'd suggest putting this at the beginning of the function. Without a user, this endpoint does nothing (as I understand it).

Reply:
I appreciate the comments, but know that I am just copying in the code from the current implementation and trying to do the minimal amount of work to get it running as quickly as possible.  There is a tradeoff between perfection and all of the other projects on my plate.  I honestly don't have time to redesign every endpoint this summer, although I realize that now is an ideal time to do so.

That said, I am not sure this should even be an API endpoint anymore.  With pages dynamically generated we can update these tables from the serve_page function in books.py when the page request is made.  It will slow down page serving ever so slightly but it eliminates an api call as well!  Seems like a reasonable tradeoff to me... thoughts??

Reply:
I like this tradeoff -- remove an endpoint, plus remove JS that calls the endpoint and does processing, and put all the code in one place. One of the tough things about client/server development is the necessity of (effectively) scattering a single unit of code across both client and server. I'm all for putting it in one place.